Timing out long running transaction

poornachandra opened the pull request:
https://issues.cask.co/browse/TEPHRA-64

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/TransactionManager.java:
```diff
@@ -325,6 +328,10 @@ private void cleanupTimedOutTransactions() {
             timedOut.add(tx.getKey());
             LOG.info("Tx invalid list: added tx {} because of timeout", tx.getKey());
+          } else if (expiration < 0) {
+            LOG.warn("Transaction {} has negative expiration time {}. " +
```
Would it be better in this case to always expire the transaction?  Having a transaction sit around forever doesn't seem good.

If it's actually due to a problem in migration, we might actually want to log a fatal error and abort the service so that we are not running with invalid data.  Though the problem with that is we need a way of clearing the state so startup can succeed.  So based on that I think it might be better to just expire the transaction.

poornachandra replied:
Agreed, will expire the transaction with a warning.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/TransactionManager.java:
```diff
@@ -1062,8 +1114,12 @@ public long getExpiration() {
       return expiration;
     }
 
+    public TransactionType getType() {
```
Annotate with @Nullable?

poornachandra replied:
Done

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/persist/TransactionEdit.java:
```diff
@@ -147,60 +161,78 @@ public static TransactionEdit createCommitting(long writePointer, Set<ChangeId>
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
I think class equality checks are inflexible, hence the priority instanceof instead.  Not a big deal though.

poornachandra replied:
Unless the derived class also makes the same instanceof checks, the equals method might become non-symmetric when instanceof is used, right?

In our case, TransactionEditV2 does not override equals method, hence instanceof will work. I will make equals final, and use insanceof check.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/persist/TransactionEdit.java:
```diff
@@ -331,6 +367,78 @@ public void encode(TransactionEdit src, DataOutput out) throws IOException {
+      dest.writePointer = in.readLong();
+      int stateIdx = in.readInt();
+      try {
+        dest.state = TransactionEdit.State.values()[stateIdx];
```
Using enum ordinals for serialization is brittle, as you have to be careful to only add new enum values at the end of the list.

I prefer using a short code (like a single byte) for each enum value and using that to (de)serialize.  I can describe in more detail if you want.

ghelmling replied:
Shouldn't this be added to the end of fields as well, instead of in the middle?

ghelmling replied:
Nevermind, misread this, thinking it was the transaction types.

ghelmling replied:
For TransactionType, I would have suggested using the byte code approach, but I don't think it makes sense to represent each of the enums differently.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/snapshot/DefaultSnapshotCodec.java:
```diff
@@ -38,8 +38,9 @@
+ * and its elements to {@code byte[]}. This codec is now deprecated and is replaced by {@link SnapshotCodecV2}.
  */
+@Deprecated
 public class DefaultSnapshotCodec implements SnapshotCodec {
```
Should this be renamed to SnapshotCodecV1?

poornachandra replied:
I didn't want to change existing class name since it is part of configuration value `data.tx.snapshot.codecs`

ghelmling commented on tephra-core/src/test/java/co/cask/tephra/TransactionServiceMainTest.java:
```diff
@@ -58,6 +60,8 @@ public void run() {
       try {
         t.start();
+        // Wait for service to startup
+        TimeUnit.SECONDS.sleep(5);
```
Why is the sleep necessary?  If there's a race, would be better to use an explicit signal or count down latch so we don't have to always wait 5 seconds.

poornachandra replied:
Fixed it by using count down latch. Also, TransactionServiceMain's start method was not returning after start-up. Changed it to return after start method is done, to comply with [Apache Commons Daemon spec] (http://commons.apache.org/proper/commons-daemon/apidocs/org/apache/commons/daemon/Daemon.html#start%28%29).

ghelmling commented on tephra-core/src/test/java/co/cask/tephra/persist/LocalTransactionStateStorageTest.java:
```diff
@@ -49,4 +66,99 @@ protected Configuration getConfiguration(String testName) throws IOException {
+        Assert.assertEquals(0, snapshot1.getInvalid().size());
+
+        // sleep longer than the cleanup interval, so that txns get timed out
+        TimeUnit.SECONDS.sleep(5);
```
Do we need to verify the timeout here?  It's already tested elsewhere and testing it is going to be very race prone.

poornachandra replied:
You are right, just verifying the following should be enough:
* Long running txn - expiry time >= TransactionManager start time + long txn timeout
* Short txn - expiry time == expiry time given during creation

I will remove the wait for timeout to happen.

ghelmling commented on tephra-core/src/test/java/co/cask/tephra/persist/LocalTransactionStateStorageTest.java:
```diff
@@ -49,4 +66,99 @@ protected Configuration getConfiguration(String testName) throws IOException {
+
+      // Start transaction manager
+      conf.setInt(TxConstants.Manager.CFG_TX_CLEANUP_INTERVAL, 3);
+      conf.setInt(TxConstants.Manager.CFG_TX_TIMEOUT, 2);
```
I don't think we should adjust the timeout values and just leave out the timeout portion of the test.

This can be done better after we port over TimestampOracle, which allows you to deterministically control the observed time.

ghelmling commented:
Mostly looks good.  A few comments though.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/TxConstants.java:
```diff
@@ -83,6 +86,13 @@
+     * The timeout for a long running transaction, in seconds. If the transaction is not finished in that time,
+     * it is marked invalid.
+     */
+    public static final String CFG_TX_LONG_TIMEOUT = "data.tx.long.timeout";
```
Please add this along with a description to the table of config properties in README.rst.

poornachandra replied:
Done.

poornachandra commented:
@ghelmling Addressed comments, please take a look.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/TransactionManager.java:
```diff
@@ -470,12 +479,34 @@ private void restoreSnapshot(TransactionSnapshot snapshot) {
+    for (Map.Entry<Long, InProgressTx> entry : inProgress.entrySet()) {
+      if (entry.getValue().getExpiration() < 0) {
+        InProgressTx compatTx =
+          new InProgressTx(entry.getValue().getVisibilityUpperBound(), getTxExpiration(defaultLongTimeout), 
```
It doesn't seem right that this is going to give a different expiration time depending on when the snapshot is being restored.  Essentially the expiration is being extended/reset on snapshot restore.  We should base the system time for the expiration on the transaction write pointer (time in millis = write pointer / TxConstants.MAX_TX_PER_MS).

poornachandra replied:
Makes sense, will derive the expiration from write pointer.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/TransactionManager.java:
```diff
@@ -516,8 +547,19 @@ private void replayLogs(Collection<TransactionLog> logs) {
+              // long running transactions were represented with expiration time as -1.
+              // This can be removed when we stop supporting TransactionEditCodecV2.
+              if (expiration < 0) {
+                expiration = getTxExpiration(defaultLongTimeout);
```
Same comment here on the expiration time.  The expiration should be based on the transaction, not when the WAL was replayed.

ghelmling commented on tephra-core/src/main/java/co/cask/tephra/snapshot/DefaultSnapshotCodec.java:
```diff
@@ -38,8 +38,9 @@
 /**
  * Handles serialization/deserialization of a {@link co.cask.tephra.persist.TransactionSnapshot}
- * and its elements to {@code byte[]}.
+ * and its elements to {@code byte[]}. This codec is now deprecated and is replaced by {@link SnapshotCodecV2}.
```
nit: you can make the extra sentence here into a \@deprecated javadoc tag

ghelmling commented on tephra-core/src/test/java/co/cask/tephra/TransactionManagerTest.java:
```diff
@@ -138,4 +138,45 @@ public void testTransactionCleanup() throws Exception {
+        // expected
+      }
+      
+      txm.abort(tx1);
```
This test made me look at the TransactionManager.doAbort() handling, and it looks like we'll have a change in behavior there due to this patch.

Previously, long running transactions would always be in progress since they weren't timed out, so we would always hit the first branch in that method -- resulting in the long running transaction going into the invalid list and staying there.  Now, with this change, if the long running transaction is timed out, and still calls abort(), it would hit the second branch in the conditional logic, resulting in the transaction ID being removed from the invalid list.  If we are making the assumption that the long running transaction's edits cannot be rolled back, then this is not what we want.  I think we should preserve the current behavior if we are relying on that in CDAP.

poornachandra replied:
Good catch! Will preserve the current behavior and will also add a test case for it.

poornachandra replied:
If `abort` is called on long running transaction, then the transaction is effectively invalidated, right? In that case, would it make sense to call `doInvalidate` from `abort` for long running transactions? This way we won't be duplicating invalidation logic in `doAbort` method, also this will simplify the conditional logic needed in `doAbort`.

ghelmling commented:
A few more comments to address here.

poornachandra commented:
@ghelmling Addressed comments, please take a look. 

I did a small re-factor to eliminate duplicate code for conversion of Transaction <-> TTransaction. That touched a few more files and increased the number of files changed.

ghelmling commented:
lgtm.

Can we squash these to a single commit with the commit message containing the JIRA?  ie. "TEPHRA-64 Time out long running transactions..."

poornachandra commented:
Thanks for the review Gary. I have squashed all the commits into https://github.com/caskdata/tephra/commit/ab139180d5f9ff88809d8573523805aa10abb150.
